### ISSUE TITLE
Blazor/Razor: Binded value constraint does not work as expected.

### ISSUE THREAD

**Issue by jshovan** [dotnet/aspnetcore#19788]
### Describe the bug
When the bound input value is updated to be equal to the previous value then subsequent updates do not behave as expected.

### To Reproduce
Copy the following into a .razor page. Then view it in a browser and update the top field. Notice that the other bound fields update as expected.
Then, update one of the bottom fields and notice that the update only acts as expected with the first character input. After the first character, the binding behavior is not as expected.

```@page "/test"

<div style="border: 5px solid black; margin: 5px;">
  <p>
    Updating the value of this input works as expected.
  </p>
  <p>
    <input value="@value" @oninput="@((ChangeEventArgs e) => value = e.Value.ToString())" />
  </p>
</div>
<div style="border: 2px solid black; margin: 5px;">
  <p>
    These do not work as expected.<br />
    When you type multiple characters the value should stay "constant" but it does not.<br />
    These update the value to "constant" with the first character input; but subsequent<br />
    characters are appended to the end of the current input and other bindings are not updated<br/>
    as demonstrated with the above input field not getting updated.
  </p>
  <p>
    <input value="@value" @oninput="@((ChangeEventArgs e) => ChangeValue(e))" />
    <input value="@value" @oninput="@((ChangeEventArgs e) => value = "Constant")" />
  </p>

</div>

@code {
  string value = "Default";

  private void ChangeValue(ChangeEventArgs e)
  {
    value = "Constant";
  }
}
```

### Further technical details

**dotnet --info**
.NET Core SDK (reflecting any global.json):
 Version:   3.1.200-preview-015002
 Commit:    b3dae24c3b

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.200-preview-015002\

Host (useful for support):
  Version: 3.1.2
  Commit:  916b5cba26

.NET Core SDKs installed:
  3.1.101 [C:\Program Files\dotnet\sdk]
  3.1.102 [C:\Program Files\dotnet\sdk]
  3.1.200-preview-015002 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  
**The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version**
Visual Studio: Version 16.5.0 Preview 5.0

**Comment by pranavkm** [comment]
> When the bound input value is updated to be equal to the previous value then subsequent updates do not behave as expected.

@jshovan can you elaborate on how you get in to situation? The examples here are fairly contrived.

**Comment by jshovan** [comment]
Notice:
1. All 3 fields are bound to the same value. So, an update to any field should update all of them. 
2. the two bottom fields are coded so that when either one of them is updated it should set value = "Constant"

------------------------------------------
1. Update the value of the top field. Observer that everything seems to work correctly. When you update the field the other 2 fields are in sync with your updates.
2. In the bottom left field type a single character "a" at the end of its existing value, to make "Constanta". 
   - notice that all 3 fields update to the value of "Constant" as expected.
3. Type another "a" character at the end of the string "Constant" in the bottom left field, to make "Constanta" again. 
  - notice that the character you typed becomes visible at the end of the string making it "Constanta". **This is incorrect. It is coded to always set the value = "Constant"**
3. Now update the bottom right field.
  - notice that updates to this field do not update the binding at all **This is wrong as it is coded to always set value = "Constant"

Now try in reverse order
4. Update the top field. Everything works as expected and the bottom fields match the value in the top field.
2. In the bottom right field type a single character "a" at the end of its existing value, to make "Constanta". 
   - notice that all 3 fields update to the value of "Constant" as expected.
3. Type another "a" character at the end of the string "Constant" in the bottom right field, to make "Constanta". 
  - notice that the character you typed becomes visible at the end of the string making it "Constanta". **This is incorrect. It is coded to always set the value = "Constant"**
3. Now update the bottom left field.
  - notice that updates to this field do not update the binding at all **This is wrong as it is coded to always set value = "Constant"

**Comment by pranavkm** [comment]
My question was more about why would have a scenario where you want to reset the value on an `onchange` event. It seems a bit unusual to have something like this in your code.

**Comment by jshovan** [comment]
I agree that it is unusual to have something like this in your code. This is just a sample that I created to point out the bug.

**Comment by pranavkm** [comment]
Fair enough. While I agree it's a bug, I'm going to park it in the backlog since it's not blocking \ impacting your application.

**Comment by jshovan** [comment]
I disagree. Consider the following example where you want to prohibit the input of certain characters. You are not able to do so. 
With this example, you should not be able to enter the values 'a' or 'b' in the bottom fields but still can because of this bug.
This is a valid case and is how I found this bug.

@page "/"
  @using System.Collections.Generic;
@using System.Linq;

  <div style="border: 5px solid black; margin: 5px;">
    <p>
      Updating the value of this input works as expected.
    </p>
    <p>
      <input value="@value" @oninput="@((ChangeEventArgs e) => value = e.Value.ToString())" />
    </p>
  </div>
  <div style="border: 2px solid black; margin: 5px;">
    <p>
      These do not work as expected.<br />
      When you type multiple characters the value should stay "constant" but it does not.<br />
      These update the value to "constant" with the first character input; but subsequent<br />
      characters are appended to the end of the current input and other bindings are not updated<br />
      as demonstrated with the above input field not getting updated.
    </p>
    <p>
      <input value="@value" @oninput="@((ChangeEventArgs e) => ChangeValue(e))" />
      <input value="@value" @oninput="@((ChangeEventArgs e) => value = "Constant")" />
    </p>

  </div>

  @code {
    char[] invalidCharacters = new char[] { 'a', 'b' };
    string value = "Default";

    private void ChangeValue(ChangeEventArgs e)
    {
      var strValue = (string) e.Value;
      if (invalidCharacters.Contains(strValue.Last()))
      {
        value = strValue.TrimEnd(invalidCharacters);
        return;
      }
      value = strValue;
    }
  }

**Comment by javiercn** [comment]
Closing since the issue is not for any currently supported version.

If you still have problems with a currently supported version, please file a new issue.
